Add tooltip component

davwheat opened the pull request:
<!--
IMPORTANT: We applaud pull requests, they excite us every single time. As we have an obligation to maintain a healthy code standard and quality, we take sufficient time for reviews. Please do create a separate pull request per change/issue/feature; we will ask you to split bundled pull requests.
-->

**Prevents breaking changes as a result of flarum/framework#2697. Progresses flarum/issue-archive#179.**

**Changes proposed in this pull request:**
This is still WIP. Created as a draft now to allow for any comments to be made while the component is still being made.

This adds a new Tooltip component to core. This should be used by extensions instead of `$.tooltip` to aid in the shift to other tooltip libraries (hopefully CSS tooltips!) in the future.

I've tried to include the main tooltip options that are used by extensions, such as `delay` and `html`, but these won't be easy or possible to keep if we shift to CSS tooltips, so they've been marked as deprecated and only exist to allow extensions to more easily shift to this component.

**Reviewers should focus on:**
Will the deprecation of `html` and `delay` cause an issue with flarum/framework#2697 as they aren't available in that? Do we consider removing deprecated features within a stable release ok, bearing in mind there is a fairly bold warning in the docblock, or should we just not implement them as part of the component in the first place.

I think implementing them as deprecated features will help ext devs migrate to this component faster and more easily now, bearing in mind the large number of other changes they need to perform.

Should we also add a deprecation warning to the tooltip helper when it's called from outside the `Tooltip` component to make this more obvious to devs?

**Screenshot**
<!-- include an image of the most relevant user-facing change, if any -->

**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.
- [ ] Backend changes: tests are green (run `composer test`).

**Required changes:**

- [ ] Related documentation PR: (Remove if irrelevant)
- [ ] Related core extension PRs: (Remove if irrelevant)

davwheat committed:
Add Tooltip component to common

Will be used to provide backwards compatibility when we switch to CSS tooltips.

All other methods of creating tooltips are deprecated and this component-based method should be used instead.

davwheat committed:
Export Tooltip component

davwheat committed:
WIP

davwheat committed:
Fix import

davwheat committed:
Add missing import

davwheat committed:
Move Tooltip jQuery helper types to separate typing file

davwheat committed:
Add `delay` attr, update docblocks

davwheat committed:
Remove undeeded `index` on import

davwheat committed:
Add support for `destroy` option in tooltip function typings

davwheat committed:
Update type import

davwheat committed:
Recreate tooltip inside `onupdate` if `text` attr has changed

davwheat committed:
Rewrite Badge to use Tooltip component; remove console.logs

davwheat committed:
Simplify Tooltip component to use `title` attr

davwheat committed:
Run `extractText` inside Tooltip component

davwheat committed:
Use `inline-block` instead of `<span>` for inline Tooltip component

davwheat committed:
Use new Tooltip component for Preview post button

davwheat committed:
Use new Tooltip component in TextEditorButton

davwheat committed:
Use new Tooltip component for DiscussionListItem-author

davwheat committed:
Allow block, inline-block, and inline container options

davwheat committed:
Use `inline-block` for TextEditorButton

davwheat committed:
Add warning when supplying `title` attr to Tooltip component

davwheat committed:
Abstract away tooltip text auto-updating logic into Tooltip component

davwheat committed:
Use new Tooltip component in PostEdited component

davwheat committed:
Add deprecation warning when `$.tooltip` is used outside of the Tooltip component

davwheat committed:
Fix broken tooltip on Badges

davwheat committed:
Explicitly state default tooltip options

davwheat commented:
I've forgotten to implement `tooltipVisible` 🙈 

I'll do that tomorrow.

davwheat committed:
WIP manual show tooltip attr

davwheat committed:
Fix deprecation warning

davwheat committed:
Add support for tooltip visibility attr

davwheat committed:
Fix prettier

askvortsov1 reviewed (commented):
The more I look at this, the less I like that the tooltip is created as an additional HTML node. Adding a tooltip is a **modification** to an existing element, not the creation of a new element. 

Would pseudoelement conflicts arise when a JS tooltip is used? Conceptually a CSS-only tooltip is cool, but the more I see workarounds needed to implement it, the less I'm convinced that CSS is ready.

SychO9 reviewed (commented):
I'm bad at typescript so I'd rather not review this, but I do have a comment

davwheat committed:
Call `super.onupdate`

davwheat committed:
Add attr to disable `title` attr warning

davwheat committed:
Modify direct child instead of using container element

Instead of using a container to house the tooltip, we'll now modify the
first direct child of the Tooltip component.

The Tooltip component will ensure that:
- children are passed to it
- only one child is present
- that child is an actual HTML Element and not a text node, or similar
- that child is currently present in the DOM

Only after all of the above are satisfied, will the tooltip be created
on that element. We store a reference to the DOM node that the tooltip
should be created on, then use this to perform tooltip actions via
jQuery. If this element gets changes (e.g. the tooltip content is
updated to another element) then the tooltip will be recreated.

If any of the first 3 requirements are not satisfied, an error will
be thrown to alert the developer to their misuse of this component.

To make this work, we do need to overwrite the title attribute of
the element with the tooltip, but this is the only solution other than
specifying `title` as an option when making the tooltip, but this is
not accessible by screenreaders unless they simulate a hover on the
element.

davwheat committed:
Use recreated Tooltip component in Badge

davwheat committed:
Add warning about component overwriting `title` attr

davwheat committed:
Remove unused styles

davwheat committed:
Update previous uses of Tooltip component

askvortsov1 reviewed (changes requested):
Overall, I like this very much! Left some relatively minor style comments.

davwheat committed:
Refactor out duplicate logic
